### PR TITLE
Added recover condition from panic on sampling policies #21903

### PR DESCRIPTION
**Pull request by narasimharaojm**
**Description:** 
Fixing the bug. It was observed that otelcollector was getting crashed with [nil p](panic: runtime error: invalid memory address or nil pointer dereference) in tail sampling processor while evaluating sampling policies. Due to this, its causing back pressure on to loadbalancing exporter.

In order to not to crash the application, added recover function to capture any nil pointer panic's and continue while evaluating sampling policies in the processor and append the count to policy evaluateErrorCount

**Link to tracking Issue:** 
#21903
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21903

**Testing:** 
Deployed the binary in my testing environment where tail sampling processing cluster evaluates the spans of size ~8M per minute and ~2.5M traces per minute. As per the added defer recover condition, I have observed increase in count of policy evaluateErrorCount as expected and application has not crashed since.

And no back pressure was observed on loadbalancing exporter
Screen Shot 2023-05-17 at 9 24 09 PM

**Documentation:** 
N/A

### CODE REVIEW

**Comment by atoulme** [comment]
Can you please add a test to reproduce the bug you are fixing here? In all likelihood, a recover call is not going to be an acceptable solution, it would be best to code defensively to avoid the panic in the first place.

Please also accept the CLA and add a changelog.

File: processor/tailsamplingprocessor/processor.go [hunk 1]
```diff
@@ -184,21 +189,32 @@ func (tsp *tailSamplingSpanProcessor) samplingPolicyOnTick() {
...
+				}
+			}()
 
-		// Sampled or not, remove the batches
-		trace.Lock()
-		allSpans := trace.ReceivedBatches
-		trace.FinalDecision = decision
-		trace.ReceivedBatches = ptrace.NewTraces()
-		trace.Unlock()
+			trace := d.(*sampling.TraceData)
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/processor.go:202]: Is this the line causing the nil pointer panic? If so, a simple nil check would work, wouldn't it?

**Comment by dmitryax** [comment]
I'm not sure the reported crash is actually a policy evaluation error. It comes from an access to a `ResourceMetrics` by an index that is not there anymore. So it sounds like some data race issue. 

@narasimharaojm could you please post the exception output with stack traces from all the goroutines? wrapped in ``` ``` here or in [the issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21903)? I'd like to investigate it before applying this fix. Ideally, we should avoid `recover()` as it's not considered a good practice in go.

File: processor/tailsamplingprocessor/processor.go (hunk 1 above)
Comment (narasimharaojm) [reply processor/tailsamplingprocessor/processor.go:202]: @jpkrohling This is the panic I observed, 

Eventually thats the line I believe its erroring out. Not sure if its due to trace data being empty or unavailability of required resource spans when trace ID fetched from idToTrace map.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x35a0519]

goroutine 241 [running]:
go.opentelemetry.io/collector/pdata/ptrace.ResourceSpans.Resource(...)
go.opentelemetry.io/collector/pdata@v1.0.0-rcv0011/ptrace/generated_resourcespans.go:58
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling.hasResourceOrSpanWithCondition({0x65f9b01?}, 0xc000b50a60, 0xc000b50a78?)
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor@v0.76.3/internal/sampling/util.go:32 +0x59
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling.(*stringAttributeFilter).Evaluate(0xc000b58750, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor@v0.76.3/internal/sampling/string_tag_filter.go:135 +0x130
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling.(*And).Evaluate(0xc000000002?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor@v0.76.3/internal/sampling/and.go:44 +0x6d
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor.(*tailSamplingSpanProcessor).makeDecision(0xc000e15860, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor@v0.76.3/processor.go:230 +0x1c4
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor.(*tailSamplingSpanProcessor).samplingPolicyOnTick(0xc000e15860)
github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor@v0.76.3/processor.go:187 +0x1a9
github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils.(*PolicyTicker).OnTick(...)
github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal@v0.76.3/timeutils/ticker_helper.go:56
github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils.(*PolicyTicker).Start.func1()
github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal@v0.76.3/timeutils/ticker_helper.go:47 +0x2e
created by github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils.(*PolicyTicker).Start
github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal@v0.76.3/timeutils/ticker_helper.go:43 +0xb0

**Pull request closed without merging**